File: src/test/java/no/nav/veilarbvedtaksstotte/client/VeilarbpersonClientImplTest.java
package no.nav.veilarbvedtaksstotte.client;

import com.github.tomakehurst.wiremock.junit.WireMockRule;
import no.nav.veilarbvedtaksstotte.utils.JsonUtils;
import no.nav.veilarbvedtaksstotte.utils.TestUtils;
import org.junit.Rule;
import org.junit.Test;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static org.junit.Assert.assertEquals;

public class VeilarbpersonClientImplTest {

    @Rule
    public WireMockRule wireMockRule = new WireMockRule(0);

    @Test
    public void skal_hente_cv_jobbprofil_json() {
        String cvJobbprofilJson = TestUtils.readTestResourceFile("cv-jobbprofil.json");
        String apiUrl = "http://localhost:" + wireMockRule.port();
        VeilarbpersonClientImpl personClient = new VeilarbpersonClientImpl(apiUrl, () -> "TOKEN");

        givenThat(get("/api/person/cv_jobbprofil?fnr=1234")
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Authorization", "Bearer TOKEN")
                        .withBody(cvJobbprofilJson))
        );

        String jsonResponse = personClient.hentCVOgJobbprofil("1234");

        assertEquals(cvJobbprofilJson, jsonResponse);
    }

    @Test
    public void skal_returnere_no_data_json_for_403_and_401() {
        String apiUrl = "http://localhost:" + wireMockRule.port();
        VeilarbpersonClientImpl personClient = new VeilarbpersonClientImpl(apiUrl, () -> "TOKEN");

        String expectedJsonResponse = JsonUtils.createNoDataStr("Bruker har ikke delt CV/jobbprofil med NAV");

        givenThat(get("/api/person/cv_jobbprofil?fnr=1234").willReturn(aResponse().withStatus(401)));
        assertEquals(expectedJsonResponse, personClient.hentCVOgJobbprofil("1234"));

        givenThat(get("/api/person/cv_jobbprofil?fnr=1234").willReturn(aResponse().withStatus(403)));
        assertEquals(expectedJsonResponse, personClient.hentCVOgJobbprofil("1234"));
    }

    @Test
    public void skal_returnere_no_data_json_for_204_and_404() {
        String apiUrl = "http://localhost:" + wireMockRule.port();
        VeilarbpersonClientImpl personClient = new VeilarbpersonClientImpl(apiUrl, () -> "TOKEN");

        String expectedJsonResponse = JsonUtils.createNoDataStr("Bruker har ikke fylt ut CV/jobbprofil");

        givenThat(get("/api/person/cv_jobbprofil?fnr=1234").willReturn(aResponse().withStatus(204)));
        assertEquals(expectedJsonResponse, personClient.hentCVOgJobbprofil("1234"));

        givenThat(get("/api/person/cv_jobbprofil?fnr=1234").willReturn(aResponse().withStatus(404)));
        assertEquals(expectedJsonResponse, personClient.hentCVOgJobbprofil("1234"));
    }

    @Test
    public void skal_sjekke_helse() {
        String apiUrl = "http://localhost:" + wireMockRule.port();
        VeilarbpersonClientImpl personClient = new VeilarbpersonClientImpl(apiUrl, () -> "TOKEN");

        givenThat(get("/internal/isAlive").willReturn(aResponse().withStatus(200)));

        personClient.checkHealth();
    }


}

File: src/test/java/no/nav/veilarbvedtaksstotte/client/person/VeilarbpersonClientImplTest.kt
package no.nav.veilarbvedtaksstotte.client.person

import com.github.tomakehurst.wiremock.client.WireMock
import com.github.tomakehurst.wiremock.junit5.WireMockRuntimeInfo
import com.github.tomakehurst.wiremock.junit5.WireMockTest
import no.nav.common.types.identer.Fnr
import no.nav.veilarbvedtaksstotte.client.person.dto.CvDto
import no.nav.veilarbvedtaksstotte.client.person.dto.CvErrorStatus
import no.nav.veilarbvedtaksstotte.client.person.dto.CvInnhold
import no.nav.veilarbvedtaksstotte.client.person.dto.PersonNavn
import no.nav.veilarbvedtaksstotte.domain.Malform
import no.nav.veilarbvedtaksstotte.utils.JsonUtils
import no.nav.veilarbvedtaksstotte.utils.TestData.TEST_FNR
import no.nav.veilarbvedtaksstotte.utils.TestUtils
import no.nav.veilarbvedtaksstotte.utils.TestUtils.givenWiremockOkJsonResponseForPost
import org.junit.Assert.assertEquals
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.fail

@WireMockTest
class VeilarbpersonClientImplTest {

    companion object {
        lateinit var veilarbpersonClient: VeilarbpersonClient

        @BeforeAll
        @JvmStatic
        fun setup(wireMockRuntimeInfo: WireMockRuntimeInfo) {
            veilarbpersonClient = VeilarbpersonClientImpl("http://localhost:" + wireMockRuntimeInfo.httpPort, {""},{""})
        }
    }

    @Test
    fun skal_hente_person() {
        WireMock.givenThat(
            WireMock.post("/api/v3/person/hent-navn")
                .withRequestBody(WireMock.equalToJson("{\"fnr\":\"$TEST_FNR\", \"behandlingsnummer\": \"" + BehandlingsNummer.VEDTAKSTOTTE.value + "\"}"))
                .willReturn(
                    WireMock.aResponse().withStatus(200).withHeader("Authorization", "Bearer TOKEN").withBody(
                        """
                               {
                                "fornavn": "Fornavn",
                                "mellomnavn": "Mellomnavn",
                                "etternavn": "Etternavn",
                                "forkortetNavn": "Sammensatt Navn"
                               }
                            """
                    )
                )
        )
        val hentPersonNavn = veilarbpersonClient.hentPersonNavn(TEST_FNR.get())
        assertEquals(
            PersonNavn(
                fornavn = "Fornavn",
                mellomnavn = "Mellomnavn",
                etternavn = "Etternavn",
                forkortetNavn = "Sammensatt Navn"
            ), hentPersonNavn
        )
    }

    @Test
    fun skal_hente_cv_jobbprofil_json() {
        val cvJobbprofilJson = TestUtils.readTestResourceFile("testdata/cv-jobbprofil.json")
        WireMock.givenThat(
            WireMock.post("/api/v3/person/hent-cv_jobbprofil")
                .withRequestBody(WireMock.equalToJson("{\"fnr\":\"1234\", \"behandlingsnummer\": \"" + BehandlingsNummer.VEDTAKSTOTTE.value + "\"}"))
                .willReturn(
                    WireMock.aResponse().withStatus(200).withHeader("Authorization", "Bearer TOKEN")
                        .withBody(cvJobbprofilJson)
                )
        )
        val cvDto = veilarbpersonClient.hentCVOgJobbprofil("1234")
        when (cvDto) {
            is CvDto.CVMedInnhold -> {
                Assertions.assertEquals(
                    JsonUtils.fromJson(cvJobbprofilJson, CvInnhold::class.java),
                    cvDto.cvInnhold
                )
            }

            is CvDto.CvMedError -> {
                fail("Unexpected error")
            }

            else -> fail("Unexpected error")
        }

    }

    @Test
    fun skal_returnere_no_data_json_for_403_og_401() {
        WireMock.givenThat(
            WireMock.post("/api/v3/person/hent-cv_jobbprofil")
                .withRequestBody(WireMock.equalToJson("{\"fnr\":\"1234\", \"behandlingsnummer\": \"" + BehandlingsNummer.VEDTAKSTOTTE.value + "\"}"))
                .willReturn(WireMock.aResponse().withStatus(401))
        )

        var cvDto = veilarbpersonClient.hentCVOgJobbprofil("1234")
        when (cvDto) {
            is CvDto.CVMedInnhold -> {
                fail("Unexpected error")
            }

            is CvDto.CvMedError -> {
                Assertions.assertEquals(cvDto.cvErrorStatus, CvErrorStatus.IKKE_DELT)
            }

            else -> fail("Unexpected error")
        }


        WireMock.givenThat(
            WireMock.post("/api/v3/person/hent-cv_jobbprofil")
                .withRequestBody(WireMock.equalToJson("{\"fnr\":\"1234\", \"behandlingsnummer\": \"" + BehandlingsNummer.VEDTAKSTOTTE.value + "\"}"))
                .willReturn(WireMock.aResponse().withStatus(403))
        )
        cvDto = veilarbpersonClient.hentCVOgJobbprofil("1234")
        when (cvDto) {
            is CvDto.CVMedInnhold -> {
                fail("Unexpected error")
            }

            is CvDto.CvMedError -> {
                Assertions.assertEquals(cvDto.cvErrorStatus, CvErrorStatus.IKKE_DELT)
            }

            else -> fail("Unexpected error")
        }
    }

    @Test
    fun skal_returnere_no_data_json_for_204_og_404() {
        WireMock.givenThat(
            WireMock.post("/api/v3/person/hent-cv_jobbprofil")
                .withRequestBody(WireMock.equalToJson("{\"fnr\":\"1234\", \"behandlingsnummer\": \"" + BehandlingsNummer.VEDTAKSTOTTE.value + "\"}"))
                .willReturn(WireMock.aResponse().withStatus(204))
        )
        var cvDto = veilarbpersonClient.hentCVOgJobbprofil("1234")
        when (cvDto) {
            is CvDto.CVMedInnhold -> {
                fail("Unexpected error")
            }

            is CvDto.CvMedError -> {
                Assertions.assertEquals(cvDto.cvErrorStatus, CvErrorStatus.IKKE_FYLT_UT)
            }

            else -> fail("Unexpected error")
        }

        WireMock.givenThat(
            WireMock.post("/api/v3/person/hent-cv_jobbprofil")
                .withRequestBody(WireMock.equalToJson("{\"fnr\":\"1234\", \"behandlingsnummer\": \"" + BehandlingsNummer.VEDTAKSTOTTE.value + "\"}"))
                .willReturn(WireMock.aResponse().withStatus(404))
        )
        cvDto = veilarbpersonClient.hentCVOgJobbprofil("1234")
        when (cvDto) {
            is CvDto.CVMedInnhold -> {
                fail("Unexpected error")
            }

            is CvDto.CvMedError -> {
                Assertions.assertEquals(cvDto.cvErrorStatus, CvErrorStatus.IKKE_FYLT_UT)
            }

            else -> fail("Unexpected error")
        }
    }

    @Test
    fun skal_sjekke_helse() {
        WireMock.givenThat(WireMock.get("/internal/isAlive").willReturn(WireMock.aResponse().withStatus(200)))
        veilarbpersonClient.checkHealth()
    }

    @Test
    fun mapper_respons_riktig() {
        val fnr = Fnr("123")
        val jsonResponse = """{"malform": "NN"}"""

        givenWiremockOkJsonResponseForPost(
            "/api/v3/person/hent-malform",
            WireMock.equalToJson("{\"fnr\":\"123\", \"behandlingsnummer\": \"" + BehandlingsNummer.VEDTAKSTOTTE.value + "\"}"),
            jsonResponse
        )

        val respons = veilarbpersonClient.hentMalform(fnr)

        assertEquals(Malform.NN, respons)
    }
}
